modules/images/regenerate-existing-images/background-process/class-perflab-background-job.php

```diff
@@ -0,0 +1,403 @@
+	 *
+	 * @return string Job status.
+	 */
+	public function get_status() {
+		return (string) get_term_meta( $this->job_id, self::JOB_STATUS_META_KEY, true );
+	}
+
+	/**
+	 * Retrieves the job data.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array|null Job data.
+	 */
+	public function get_data() {
+		// If we have cached data, return it.
+		if ( ! is_null( $this->data ) ) {
+			return $this->data;
+		}
+
+		$this->data = get_term_meta( $this->job_id, self::JOB_DATA_META_KEY, true );
+
+		return $this->data;
+	}
+
+	/**
+	 * Retrieves the job name.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return string Job name.
+	 */
+	public function get_name() {
+		return (string) $this->name;
+	}
+
+	/**
+	 * Creates a new job in the queue.
+	 *
+	 * Job refers to the term and queue refers to the `background_job` taxonomy.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $name Name of job identifier.
+	 * @param array  $data Data for the job.
+	 *
+	 * @return Perflab_Background_Job|WP_Error Job object if created successfully, else WP_Error.
+	 */
+	public static function create( $name, array $data = array() ) {
+		// Insert the new job in queue.
+		$term_data = wp_insert_term( 'job_' . time(), PERFLAB_BACKGROUND_JOB_TAXONOMY_SLUG );
+
+		if ( ! is_wp_error( $term_data ) ) {
+			update_term_meta( $term_data['term_id'], self::JOB_DATA_META_KEY, $data );
+			update_term_meta( $term_data['term_id'], self::JOB_NAME_META_KEY, $name );
+
+			/**
+			 * Create a fresh instance to return.
+			 *
+			 * @var Perflab_Background_Job
+			 */
+			$job = new self( $term_data['term_id'] );
+
+			$job->name = $name;
+			$job->data = $data;
+
+			// Set the queued status for freshly created jobs.
+			$job->set_status( self::JOB_STATUS_QUEUED );
+
+			return $job;
+		}
+
+		return $term_data;
+	}
+
+	/**
+	 * Returns the number of attempts executed for a job.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return int
+	 */
+	public function get_attempts() {
+		$attempts = get_term_meta( $this->job_id, self::JOB_ATTEMPTS_META_KEY, true );
+
+		return absint( $attempts );
+	}
+
+	/**
+	 * Set the start time of job.
+	 * It tells at what point of time the job has been started.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param int $time Timestamp (in seconds) at which job has been started.
+	 *
+	 * @return void
+	 */
+	public function lock( $time = null ) {
+		$time = empty( $time ) ? time() : $time;
+		update_term_meta( $this->job_id, self::JOB_LOCK_META_KEY, $time );
+		$this->set_status( self::JOB_STATUS_RUNNING );
+	}
+
+	/**
+	 * Get the timestamp (in seconds) when the job was started.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return int Start time of the job.
+	 */
+	public function get_start_time() {
+		$time = get_term_meta( $this->job_id, self::JOB_LOCK_META_KEY, true );
+
+		return absint( $time );
+	}
+
+	/**
+	 * Unlocks the process.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return void
+	 */
+	public function unlock() {
+		delete_term_meta( $this->job_id, self::JOB_LOCK_META_KEY );
+	}
```

Comment:
It's a bit confusing and inconsistent that `lock()` also sets the job status, while `unlock()` doesn't. I'm not sure whether both should change the status or neither should change the status, but we should update them to be consistent, so that the `unlock()` truly "undoes" the `lock()`.